src/compiler/checker.ts

```diff
@@ -28040,6 +28036,27 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return getUnionType(map(clauseWitnesses, text => text ? narrowTypeByTypeName(type, text) : neverType));
         }
 
+        function narrowTypeBySwitchOnTrue(type: Type, switchStatement: SwitchStatement, clauseStart: number, clauseEnd: number): Type {
+            const clauses = switchStatement.caseBlock.clauses.slice(clauseStart, clauseEnd);
+            const defaultIndex = findIndex(switchStatement.caseBlock.clauses, clause => clause.kind === SyntaxKind.DefaultClause);
+            const hasDefaultClause = clauseStart === clauseEnd || (defaultIndex >= clauseStart && defaultIndex < clauseEnd);
+            if (hasDefaultClause) {
+                // In the default case, we need to assume that all of the other cases were false,
+                // but also that any of the other cases in this block are true.

```

Comment:
```suggestion
                // In the default case, it's possible that we came from:
                // 1. a set of fallthrough cases, where their conditions might be true
                // 2. no case succeeding, in which case all conditions might be false
```

Reply:
Actually...rereading it, is it the case that we gain no information from considering the fallthrough cases? Since you have to assume they could be both true and false?

Reply:
Yeah, that's the edge case I'm guessing at in https://github.com/microsoft/TypeScript/pull/55991#issuecomment-1749364932, I need to write a test that can show that possibility.

Not to mention that you could have a broad case and then narrow within the case and return, leaving only part of the type to fallthough...

Reply:
In thinking harder, I realized that what this function should do is just "what do I know about this thing based on only the case itself?"; you don't need to worry about control flow, fallthrough, etc, here at all, because CFA will combine this info with the other cases later.

So in reality, the bug being fixed is "multiple cases were not handled" and "default case was not handled".